### ISSUE TITLE
Support different activations for LSTM layer at ONNX importer

### ISSUE THREAD

**Issue by heiheiya** [opencv/opencv#18390]
##### System information (version)
- OpenCV => :4.4:
- Operating System / Platform => :Linux:
- Compiler => :None:

##### Detailed description
I have trained a simple lstm model use keras, ans then I use keras2onnx to convert keras h5 model to onnx model. Finally, I use opencv 4.4 readNetFromONNX to load the model, but I got the error:
/opencv/modules/dnn/src/onnx/onnx_importer.cpp: error: (-213: The function/feature is not implemented) Unexpected attribute type in function 'getLayerParams'
Below is my keras model and onnx model opened with netron.
![image](https://user-images.githubusercontent.com/8655449/93973646-20360b80-fda7-11ea-8bd7-60fab8602a94.png)

![image](https://user-images.githubusercontent.com/8655449/93973433-c3d2ec00-fda6-11ea-9e09-06f906cd52bb.png)

**Comment by alalek** [comment]
Could you please attach minimal `.onnx` file? (with random weights)
We need a reproducer to ensure and check that this problem got fixed.

**Comment by heiheiya** [comment]
@alalek here is my onnx file convert from keras h5 file.

[test.zip](https://github.com/opencv/opencv/files/5273049/test.zip)

**Comment by alalek** [comment]
@heiheiya Thank you!

LSTM is not fully supported yet in ONNX importer.

Current importer implementation fails to handle `activations` attribute: `HardSigmoid, Tanh, Tanh`.

**Comment by heiheiya** [comment]
@alalek Thank you!
But I want to know which importer in opencv can fully support LSTM? I have tried to convert keras model(unrolling LSTM) to tensorflow pb model, the tensorflow importer have failed too.
Is there any way to deploy LSTM model use opencv?

**Comment by smbz** [comment]
OpenCV currently only supports sigmoid activation for LSTM gates, not hard sigmoid.  Additionally, there seems to be a bug in the ONNX importer where it fails to handle the `activations` attribute even if the value is supported.

@heiheiya tanh and sigmoid is the default for ONNX, so it might be worth retraining with sigmoid instead of hard sigmoid and seeing if that works.

**Comment by heiheiya** [comment]
@smbz Thank you for your reply! I took your advice, changed LSTM activation tanh and hard sigmoid to sigmoid and retrained the network. Unfortunately, I've got the same error just like before(ONNX importer and tensorflow importer, both got the same error like before, but each got different error). There may be other factors contributing to the error.
I attached the new model file.
[testLSTM-sigmoid.zip](https://github.com/opencv/opencv/files/5296787/testLSTM-sigmoid.zip)

**Comment by smbz** [comment]
@heiheiya I meant sigmoid just for the gates - keep using tanh for the other activations.  In other words, sigmoid replaces the hard sigmoid in your original network, but everything else remains the same.

**Comment by heiheiya** [comment]
@smbz I have tried, also the same error, I don't know why?

**Comment by berak** [comment]
@heiheiya the activations in the LSTM layer are currently hardcoded. it's probably, because [the pytorch implementation](https://pytorch.org/docs/master/generated/torch.nn.LSTM.html#lstm) has it so, while it's configurable [in keras](https://keras.io/api/layers/recurrent_layers/lstm/)


opencv's onnx importer looks at *all* given attributes, and errors out if it finds something unknown (a list of "activations" in this case), so you're out of luck with your keras model, unless the onnx importer and the LSTM layer get fixed

**Comment by heiheiya** [comment]
@berak After so many failures, I found a way. Train the LSTM network use tf.keras not keras, (keep unroll=False), then convert h5 file to pb file use the solution proposed by [https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/util/tf_graph_util.py](url), then use cv2.dnn.readNetFromTensorflow can load the model successfully.
But now I've got a new error when use the net to forward
dnn.cpp:525: error: (-2:Unspecified error) Can't create layer "lstm/while/Enter" of type "Enter"

**Comment by aitikgupta** [comment]
@alalek Since e59793cc75ed5476c7d85a61c539c5cf8944a54c, the error is more logged, but still isn't handled. Here's the updated error log [for the onnx file @heiheiya provided](https://github.com/opencv/opencv/files/5273049/test.zip):

```
[ERROR:0] DNN/ONNX/Attribute[activations]: 'Strings' (3) are not supported
[ERROR:0]     Attribute[activations].string(0) = 'HardSigmoid'
[ERROR:0]     Attribute[activations].string(1) = 'Tanh'
[ERROR:0]     Attribute[activations].string(2) = 'Tanh'
[ERROR:0] DNN/ONNX: ERROR during processing node with 8 inputs and 3 outputs: [LSTM]:(lstm_1_Y)
...
...
error: (-213:The function/feature is not implemented) DNN/ONNX/Attribute[activations]: 'Strings' (3) are not supported in function 'getLayerParams'
```

**I can confirm the error even for `Sigmoid, Tanh, Tanh`**
I think there needs to be an additional condition for `attribute_name == "activations"` around around [here](https://github.com/opencv/opencv/blob/969b55036fd4e3764194995de9ea706f35a65c9a/modules/dnn/src/onnx/onnx_importer.cpp#L249-L250)
According to [the assertion here](https://github.com/opencv/opencv/blob/969b55036fd4e3764194995de9ea706f35a65c9a/modules/dnn/src/onnx/onnx_importer.cpp#L866) Number of inputs should be 7, but for both the onnx files the number of inputs is 8


Here's the onnx file:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/testdata/CNTK/test_LSTM.tanh.bidirectional/model.onnx

![Screenshot from 2020-10-02 05-50-55](https://user-images.githubusercontent.com/43996118/94875723-3f9dfa00-0473-11eb-933b-1664d2d94138.png)

**Comment by heiheiya** [comment]
@berak @smbz @alalek @aitikgupta Now, I got a error at net.forward() use LSTM pb model
cv2.error: OpenCV(4.4.0) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-hfoi92lm\opencv\modules\dnn\src\dnn.cpp:604: error: (-2:Unspecified error) Can't create layer "lstm/Shape_1" of type "Shape" in function 'cv::dnn::dnn4_v20200609::LayerData::getLayerInstance'
[lstm.zip](https://github.com/opencv/opencv/files/5353419/lstm.zip)

**Comment by leochan2009** [comment]
Yep, same issue here, any progress? Thanks

**Comment by heiheiya** [comment]
@leochan2009, no, maybe need official improvement.

**Comment by sl-sergei** [comment]
@heiheiya Can you please provide inference code for your original model (Keras) or ONNX? I started investigating this issue and made a fix for loading of the network, but inference results re different from ONNX Runtime results.

**Comment by sl-sergei** [comment]
> @berak @smbz @alalek @aitikgupta Now, I got a error at net.forward() use LSTM pb model
> cv2.error: OpenCV(4.4.0) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-hfoi92lm\opencv\modules\dnn\src\dnn.cpp:604: error: (-2:Unspecified error) Can't create layer "lstm/Shape_1" of type "Shape" in function 'cv::dnn::dnn4_v20200609::LayerData::getLayerInstance'
> [lstm.zip](https://github.com/opencv/opencv/files/5353419/lstm.zip)

The problem with this model (lstm.pb) is that the graph fragmented into multiple atomic operations instead of single LSTM layer:
![image](https://user-images.githubusercontent.com/62052793/105213405-b7497f00-5b5f-11eb-8340-cf8bd5683458.png)

**Comment by asmorkalov** [comment]
Merged to master.
